autopilot/agent_test.go

```diff
@@ -1699,3 +1699,390 @@ func TestAgentQuitWhenPendingConns(t *testing.T) {
+				numNodes, len(req.nodes))
+		}
+
+		// One channel is already open.
+		if len(req.chans) != 1 {
+			t.Fatalf("expected 1 existing channel, got %v",
+				len(req.chans))
+		}
+		if req.amt != chanSize {
+			t.Fatalf("expected channel size of %v, got %v",
+				chanSize, req.amt)
+		}
+
+	case <-time.After(time.Second * 3):
+		t.Fatalf("select wasn't queried in time")
+	}
+
+	// Respond with scores for the all nodes.
+	select {
+	case heuristic.nodeScoresResps <- nodeScores:
+	case <-time.After(time.Second * 3):
+		t.Fatalf("NodeScores wasn't queried in time")
+	}
+
+	// The agent should attempt to open channels, totaling its channel
+	// budget.
+	var totalAllocation btcutil.Amount
+	for i := 0; i < numChans; i++ {
+		select {
+		case openChan := <-chanController.openChanSignals:
+			totalAllocation += openChan.amt
+			if openChan.amt != chanSize {
+				t.Fatalf("invalid chan amt: expected %v, "+
+					"got %v", chanSize, openChan.amt)
+			}
+
+			delete(nodeScores, NewNodeID(openChan.target))
+
+		case <-time.After(time.Second * 3):
+			t.Fatalf("channel not opened in time")
+		}
+	}
+
+	if totalAllocation != channelBudget {
+		t.Fatalf("expected agent to open channels totalling %v, "+
+			"instead was %v", channelBudget, totalAllocation)
+	}
+
+	// TODO(halseth): this loop is a hack to ensure all the attempted
+	// channels are accounted for. This happens because the agent will
+	// query the ChannelBudget before all the pending channels are added to
+	// the map. Fix by adding them to the pending channels map before
+	// executing directives in goroutines?
+	waitForNumChans := func(expChans int) {
+		t.Helper()
+
+	Loop:
+		for {
+			select {
+			case arg := <-constraints.moreChanArgs:
+				// As long as the number of existing channels
+				// is below our expected number of channels,
+				// we'll keep responding with "no more
+				// channels".
+				if len(arg.chans) != expChans {
+					select {
+					case constraints.moreChansResps <- moreChansResp{0, 0}:
+					case <-time.After(time.Second * 3):
+						t.Fatalf("heuristic wasn't queried " +
+							"in time")
+					}
+					continue
+				}
+
+				if arg.balance != availableFunds {
+					t.Fatalf("expectd agent to have %v balance, had %v",
+						availableFunds, arg.balance)
+				}
+				break Loop
+
+			case <-time.After(time.Second * 3):
+				t.Fatalf("heuristic wasn't queried in time")
+			}
+		}
+	}
+
+	// Wait for the agent to have numChans+1 a channels.
+	waitForNumChans(numChans + 1)
+
+	// Set the channel budget to 1.5 BTC.
+	channelBudget = btcutil.SatoshiPerBitcoin * 3 / 2
+
+	// We'll return a response telling the agent to open 3 channels, with a
+	// total channel budget of 1.5 BTC.
+	select {
+	case constraints.moreChansResps <- moreChansResp{
+		numMore: 3,
+		amt:     channelBudget,
+	}:
+	case <-time.After(time.Second * 10):
+		t.Fatalf("heuristic wasn't queried in time")
+	}
+
+	// Since the maximum channel size is 1 BTC, the agent won't have enough
```

Comment:
I find these tests hard to follow at times in general. I think we can reduce some vertical space here with usage of some a helper functions to handle sending/receiving messages to/from the various mocked interfaces. I'd expect the logic to go "ok I have 1.5 BTC and the max size is 1 BTC, so I'll open to channels here for 1 and 0.5 BTC". Instead, as is, it'll instead execute the 1 BTC and the 0.5 BTC during distinct iterations. 

This is a regression compared to the behavior of the prior system, that would attempt to pipeline open as many channels as possible.

Reply:
The rationale behind doing it this way was that you don't know if an opening attempt will always succeed, so I would wait until the next iteration to have more information about which ones succeed and potentially open a bigger channel.

Looking at it now though, I see that it won't matter much, since we will eventually end up with the smaller channel anyway, so whether it is being opened last or not makes no difference. Will fix!

Also agree on the verbosity of the tests, will clean they up a bit.